### PR TITLE
Transforming S3 PUT override

### PR DESCRIPTION
**Pull request by dagnir**
This transforms the AmazonS3#putObject(String,String,File) override.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                + "    public static void main(String[] args) {\n"
+                + "        AmazonS3Client s3 = null;\n"
+                + "\n"
+                + "        File myFile = new File(\"test.txt\");\n"
+                + "\n"
+                + "        s3.putObject(BUCKET, KEY, myFile);\n"
+                + "    }\n"
+                + "}\n",
+                "import com.amazonaws.services.s3.AmazonS3Client;\n"
+                + "import com.amazonaws.services.s3.model.PutObjectRequest;\n"
```
Comment (davidh44) [inline migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:61]: Should these imports be updated to v2 or is this handled by another recipe?

File: migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                + "public class S3PutObjectExample {\n"
+                + "    private static final String BUCKET = \"my-bucket\";\n"
+                + "    private static final String KEY = \"key\";\n"
+                + "\n"
+                + "    public static void main(String[] args) {\n"
+                + "        AmazonS3Client s3 = null;\n"
+                + "\n"
+                + "        File myFile = new File(\"test.txt\");\n"
+                + "\n"
+                + "        s3.putObject(new PutObjectRequest(BUCKET, KEY), RequestBody.fromFile(myFile));\n"
```
Comment (davidh44) [inline migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:75]: Should it be builder for PutObjectRequest? What about AsyncRequestBody for async client?

File: migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new Visitor();
+    }
+
+    private static final class Visitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext executionContext) {
+            if (PUT_OBJECT.matches(method, false)) {
+                method = transformPutFileOverload(method);
```
Comment (zoewangg) [inline migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java:70]: Don't we need to invoke super.visitMethodInvocation?

File: migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+            return method.withMethodType(methodType).withArguments(newArgs);
+        }
+
+        private J.MethodInvocation fileToRequestBody(Expression fileExpr) {
+            maybeAddImport(REQUEST_BODY);
+
+            J.Identifier requestBodyId = IdentifierUtils.makeId(REQUEST_BODY.getClassName(), REQUEST_BODY);
+
+            JavaType.Method fromFileType = new JavaType.Method(
```
Comment (zoewangg) [inline migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java:112]: Those 9-argu ctors always look intimating to me 😨

File: migration-tool/src/main/resources/META-INF/rewrite/s3-putbject-constructor-to-fluent.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: software.amazon.awssdk.S3PutObjectConstructorToFluent
+displayName: Change auth related classes
```
Comment (zoewangg) [inline migration-tool/src/main/resources/META-INF/rewrite/s3-putbject-constructor-to-fluent.yml:18]: nit: display name is not updated

File: migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java (hunk 1 above)
Comment (zoewangg) [reply migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:61]: It's odd the tests are not failing. Maybe we can apply multiple recipes? That's what I did for other tests https://github.com/aws/aws-sdk-java-v2/blob/feature/master/migration-tool/migration-tool/src/test/java/software/amazon/awssdk/migration/recipe/ChangeConfigTypesTest.java#L42

File: migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java (hunk 2 above)
Comment (dagnir) [reply migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:75]: > Should it be builder for PutObjectRequest?

No, this is intentional. Changing this to the builder pattern is left to a downstream recipe https://github.com/aws/aws-sdk-java-v2/blob/e59e29d7927e8bf099a9118ee287d96ec1e5c565/migration-tool/src/main/resources/META-INF/rewrite/s3-putbject-constructor-to-fluent.yml

> What about AsyncRequestBody for async client?
There is no async client for v1 s3.

File: migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java (hunk 3 above)
Comment (dagnir) [reply migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java:70]: We do, on 72. I don't think ordering matters here

File: migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java (hunk 1 above)
Comment (zoewangg) [reply migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:61]: Oh, so the tests actually are not failing due to method reference not found (don't remember the exact message)?

Comment (dagnir) [reply migration-tool/src/test/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2Test.java:61]: Yeah. I think that only fails when loading the `before` code, where it does the type validation

File: migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java (hunk 4 above)
Comment (dagnir) [reply migration-tool/src/main/java/software/amazon/awssdk/migration/internal/recipe/S3StreamingRequestToV2.java:112]: Yeah. I think if it becomes a bigger issue we can add some util methods to take care of the boilerplate

**Review by davidh44 — APPROVED** [review]

**Review by zoewangg — APPROVED** [review]

**Comment by zoewangg** [comment]
Closing in favor of #5411

**Pull request closed without merging**